jass/lib/UnitDex.j

```diff
@@ -0,0 +1,434 @@
+*       function TriggerRegisterUnitIndexEvent takes trigger t, integer eventtype returns nothing
+*
+*       function OnUnitIndex takes code func returns triggercondition
+*       function OnUnitDeidex takes code func returns triggercondition
+*   _________________________________________________________________________
+*   4. Struct API
+*   ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
+*       UnitDex.Enabled = false // toggle the indexer
+*       UnitDex.Initialized     // returns true if the preload timer has finished
+*       UnitDex.Count           // returns the amount of units indexed
+*       UnitDex.Unit[0]         // access the UnitDex array directly
+*       UnitDex.Group           // a unit group containing every indexed unit (for enumeration)
+*       UnitDex.LastIndex       // returns the last indexed unit's id
+*   _________________________________________________________________________
+*   5. Public Variables
+*   ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
+*       These are to be used with the "eventtype" argument of the event API:
+*
+*           constant integer EVENT_UNIT_INDEX     = 0
+*           constant integer EVENT_UNIT_DEINDEX   = 1
+*   _________________________________________________________________________
+*   6. Examples
+*   ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
+*       1. Associate a unit with a variable
+*
+*           set UnitFlag[GetUnitId(yourUnit)] = true
+*
+*       2. Allocate a struct instance using a units assigned ID
+*
+*           local somestruct data = GetUnitId(yourUnit)
+*  
+*       3. Detect when a unit leaves the map
+*
+*           function Exit takes nothing returns nothing
+*               call BJDebugMsg("The unit " + GetUnitName(GetIndexedUnit()) + " has left the map.")
+*           endfunction
+*
+*           call OnUnitDeindex(function Exit)
+*           // or
+*           call RegisterUnitIndexEvent(Filter(function Exit), EVENT_UNIT_DEINDEX)
+*
+*
+*   _________________________________________________________________________
+*   7. How it works
+*   ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
+*       1. Enumerate all preplaced units
+*       2. TriggerRegisterEnterRegion native to detect when a unit enters the map
+*       3. Assign each unit that enters the map a unique integer
+*       4. Give every unit an ability based off of defend. There is no native to accurately
+*          detect when a unit leaves the map but when a unit dies or is removed from the game
+*          they are issued the "undefend" order
+*       5. Catch the "undefend" order and remove unit from the queue
+*   _________________________________________________________________________
+*   8. Notes
+*   ¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
+*       - This system is compatable with GUI because it utilizes UnitUserData (custom values for units).
```

Comment:
Unfortunately we already use `UnitUserData ` (code from before my time) so we'll need to write a shim for those cases. I believe its contained to just the Hydra functions though, thankfully.